runtime/callback.c

```diff
@@ -294,27 +293,60 @@ CAMLexport value caml_callbackN_exn(value closure, int narg, value args[]) {
 
 #endif
 
+/* Result-returning variants of the above */
+
+CAMLexport caml_result caml_callbackN_result(
+  value closure, int narg, value args[])
+{
+  return Result_encoded(caml_callbackN_exn(closure, narg, args));
+}
+
+CAMLexport caml_result caml_callback_result(
+  value closure, value arg)
+{
+  return Result_encoded(caml_callback_exn(closure, arg));
```

Comment:
The plan is to keep using the "encoded exceptions" format for code that is defined in the assembly runtime, because there it is much easier to flip a boolean than to return a structure. Now assuming that all the `caml_callback*_asm` functions return an encoded exception, it is more logical I thnink to have the `_exn` version be primitive and the `_result` version defined on top of it -- otherwise the `_exn` version would go back and forth, from encoded exceptions to caml_result to encoded exception again.